File: scripts/run_job_agents.py
#!/usr/bin/env python3
"""
Run Job Application Assistant AI Agents (Python with sandboxing).

This script starts 3 autonomous AI agents that help with job applications:
- Agent 1: Job Hunter (OpenAI) - Searches and scores job matches
- Agent 2: CV Crafter (Claude) - Generates tailored CVs and cover letters
- Agent 3: Application Tracker (Grok) - Tracks applications and generates insights

Usage:
    python scripts/run_job_agents.py
    or
    USE_SANDBOX=true python scripts/run_job_agents.py
"""

import asyncio
import os
import sys
from pathlib import Path

# Add pyagents to path
sys.path.insert(0, str(Path(__file__).parent.parent))

from pyagents.job_hunter_agent import JobHunterAgent, JobHunterConfig
from pyagents.cv_crafter_agent import CVCrafterAgent, CVCrafterConfig
from pyagents.application_tracker_agent import ApplicationTrackerAgent, ApplicationTrackerConfig
from pyagents.job_llm_providers import JobLLMProviderManager


def load_dev_vars() -> dict:
    """Load environment variables from .dev.vars file."""
    vars_dict = {}
    dev_vars_path = Path(__file__).parent.parent / ".dev.vars"
    
    if dev_vars_path.exists():
        with open(dev_vars_path, "r") as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith("#") and "=" in line:
                    key, value = line.split("=", 1)
                    vars_dict[key.strip()] = value.strip()
    
    return vars_dict


async def main():
    """Main function to run all agents."""
    # Load environment variables
    dev_vars = load_dev_vars()
    for key, value in dev_vars.items():
        if key not in os.environ:
            os.environ[key] = value

    # Configuration from environment variables
    database_url = os.environ.get("DATABASE_URL")
    target_user_id = os.environ.get("TARGET_USER_ID", "samuel_student")

    # LLM API Keys
    openai_api_key = os.environ.get("OPENAI_API_KEY")
    anthropic_api_key = os.environ.get("ANTHROPIC_API_KEY")
    grok_api_key = os.environ.get("GROK_API_KEY") or os.environ.get("XAI_API_KEY")

    # Sandbox setting
    use_sandbox = os.environ.get("USE_SANDBOX", "false").lower() == "true"

    print("💼 Job Application Assistant - Multi-Agent System (Python)")
    print("=" * 60)
    print(f"Database: {'Connected' if database_url else 'NOT SET'}")
    print(f"Target User: {target_user_id}")
    print(f"Sandbox: {'Enabled' if use_sandbox else 'Disabled'}")
    print("=" * 60 + "\n")

    if not database_url:
        print("❌ ERROR: DATABASE_URL environment variable is required")
        sys.exit(1)

    # Validate API keys
    if not openai_api_key:
        print("❌ ERROR: OPENAI_API_KEY environment variable is required")
        sys.exit(1)
    if not anthropic_api_key:
        print("❌ ERROR: ANTHROPIC_API_KEY environment variable is required")
        sys.exit(1)
    if not grok_api_key:
        print("⚠️  WARNING: GROK_API_KEY not set. Application Tracker agent will use fallback behavior.")

    # Initialize LLM manager
    llm_manager = JobLLMProviderManager(
        openai_api_key=openai_api_key,
        anthropic_api_key=anthropic_api_key,
        grok_api_key=grok_api_key,
    )

    print("🤖 Initializing agents...\n")

    # Create agent configs
    job_hunter_config = JobHunterConfig(
        name="Job Hunter",
        agent_id="job-hunter-openai",
        database_url=database_url,
        llm_provider="openai",
        llm_api_key=openai_api_key,
        llm_model="gpt-4o",
        target_user_id=target_user_id,
    )

    cv_crafter_config = CVCrafterConfig(
        name="CV Crafter",
        agent_id="cv-crafter-claude",
        database_url=database_url,
        llm_provider="anthropic",
        llm_api_key=anthropic_api_key,
        llm_model="claude-3-sonnet-20240229",
        target_user_id=target_user_id,
    )

    application_tracker_config = ApplicationTrackerConfig(
        name="Application Tracker",
        agent_id="application-tracker-grok",
        database_url=database_url,
        llm_provider="grok",
        llm_api_key=grok_api_key,
        llm_model="grok-beta",
        target_user_id=target_user_id,
    )

    # Create agents
    job_hunter = JobHunterAgent(job_hunter_config, llm_manager, use_sandbox=use_sandbox)
    cv_crafter = CVCrafterAgent(cv_crafter_config, llm_manager, use_sandbox=use_sandbox)
    application_tracker = ApplicationTrackerAgent(
        application_tracker_config, llm_manager, use_sandbox=use_sandbox
    )

    # Connect all agents
    await job_hunter.connect()
    await cv_crafter.connect()
    await application_tracker.connect()

    print("\n✅ All 3 agents initialized!")
    print("\n🚀 Starting agent iterations...\n")
    print("Press Ctrl+C to stop\n")

    iteration = 1
    try:
        while True:
            print("=" * 60)
            print(f"ITERATION {iteration}")
            print("=" * 60)

            # Run all agents in parallel
            await asyncio.gather(
                job_hunter.take_turn(iteration),
                cv_crafter.take_turn(iteration),
                application_tracker.take_turn(iteration),
            )

            print(f"\n✅ Iteration {iteration} complete!")
            print("⏰ Next iteration in 5 minutes...\n")

            # Wait 5 minutes before next iteration
            await asyncio.sleep(300)  # 5 minutes
            iteration += 1

    except KeyboardInterrupt:
        print("\n\n🛑 Stopping agents...")
    finally:
        # Disconnect all agents
        await job_hunter.disconnect()
        await cv_crafter.disconnect()
        await application_tracker.disconnect()
        print("\n✅ All agents disconnected. Goodbye!")


if __name__ == "__main__":
    # On Windows, psycopg async requires selector-based loops instead of Proactor.
    if sys.platform.startswith("win"):
        asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
    asyncio.run(main())

File: scripts/run-job-agents.ts
#!/usr/bin/env node
/**
 * Run Job Application Assistant AI Agents
 * 
 * This script starts 3 autonomous AI agents that help with job applications:
 * - Agent 1: Job Hunter (OpenAI) - Searches and scores job matches
 * - Agent 2: CV Crafter (Claude) - Generates tailored CVs and cover letters
 * - Agent 3: Application Tracker (Grok) - Tracks applications and generates insights
 * 
 * Usage:
 *   npm run agents:start
 *   or
 *   tsx scripts/run-job-agents.ts
 */

import { readFileSync } from "fs";
import { join } from "path";
import { JobHunterAgent, JobHunterConfig } from "../src/agents/job-hunter-agent.js";
import { CVCrafterAgent, CVCrafterConfig } from "../src/agents/cv-crafter-agent.js";
import { ApplicationTrackerAgent, ApplicationTrackerConfig } from "../src/agents/application-tracker-agent.js";
import { LLMProviderManager } from "../src/agents/llm-providers.js";
import { JobDatabaseClient } from "../src/agents/job-database-client.js";

// Load .dev.vars file (same format as wrangler uses)
function loadDevVars(): void {
	try {
		const devVarsPath = join(process.cwd(), ".dev.vars");
		const content = readFileSync(devVarsPath, "utf-8");
		
		for (const line of content.split("\n")) {
			const trimmed = line.trim();
			if (trimmed && !trimmed.startsWith("#") && trimmed.includes("=")) {
				const [key, ...valueParts] = trimmed.split("=");
				const value = valueParts.join("=").trim();
				if (key && value && !process.env[key]) {
					process.env[key] = value;
				}
			}
		}
	} catch (error) {
		// .dev.vars might not exist, that's okay
		console.warn("⚠️  Could not load .dev.vars, using environment variables only");
	}
}

// Load environment variables
loadDevVars();

// Configuration from environment variables
const DATABASE_URL = process.env.DATABASE_URL;
const TARGET_USER_ID = process.env.TARGET_USER_ID || "samuel_student"; // Default user

// LLM API Keys (required)
const OPENAI_API_KEY = process.env.OPENAI_API_KEY;
const ANTHROPIC_API_KEY = process.env.ANTHROPIC_API_KEY;
const GROK_API_KEY = process.env.GROK_API_KEY || process.env.XAI_API_KEY;

async function main() {
	console.log("💼 Job Application Assistant - Multi-Agent System");
	console.log("=" .repeat(60));
	console.log(`Database: ${DATABASE_URL ? "Connected" : "NOT SET"}`);
	console.log(`Target User: ${TARGET_USER_ID}`);
	console.log("=" .repeat(60) + "\n");

	if (!DATABASE_URL) {
		console.error("❌ ERROR: DATABASE_URL environment variable is required");
		process.exit(1);
	}

	// Validate API keys
	if (!OPENAI_API_KEY) {
		console.error("❌ ERROR: OPENAI_API_KEY environment variable is required");
		process.exit(1);
	}
	if (!ANTHROPIC_API_KEY) {
		console.error("❌ ERROR: ANTHROPIC_API_KEY environment variable is required");
		process.exit(1);
	}
	if (!GROK_API_KEY) {
		console.warn("⚠️  WARNING: GROK_API_KEY not set. Application Tracker agent will use fallback behavior.");
	}

	// Wait for user input from the frontend before starting iterations
	const onboardingSnapshot = await waitForUserSetup(DATABASE_URL!, TARGET_USER_ID);
	printUserSnapshot(onboardingSnapshot);

	// Initialize LLM Manager
	const llmManager = new LLMProviderManager(
		OPENAI_API_KEY,
		ANTHROPIC_API_KEY,
		GROK_API_KEY
	);

	// Define 3 agents
	const jobHunterConfig: JobHunterConfig = {
		name: "Job Hunter",
		agentId: "job-hunter-openai",
		databaseUrl: DATABASE_URL!,
		llmProvider: "openai",
		llmApiKey: OPENAI_API_KEY,
		llmModel: "gpt-4o-mini",
		targetUserId: TARGET_USER_ID,
	};

	const cvCrafterConfig: CVCrafterConfig = {
		name: "CV Crafter",
		agentId: "cv-crafter-claude",
		databaseUrl: DATABASE_URL!,
		llmProvider: "anthropic",
		llmApiKey: ANTHROPIC_API_KEY,
		llmModel: "claude-3-5-haiku-latest",
		targetUserId: TARGET_USER_ID,
	};

	const applicationTrackerConfig: ApplicationTrackerConfig = {
		name: "Application Tracker",
		agentId: "application-tracker-grok",
		databaseUrl: DATABASE_URL!,
		llmProvider: "grok",
		llmApiKey: GROK_API_KEY,
		llmModel: "grok-3",
		targetUserId: TARGET_USER_ID,
	};

	// Create agents
	console.log("🤖 Initializing agents...\n");
	const jobHunter = new JobHunterAgent(jobHunterConfig, llmManager);
	const cvCrafter = new CVCrafterAgent(cvCrafterConfig, llmManager);
	const applicationTracker = new ApplicationTrackerAgent(applicationTrackerConfig, llmManager);

	await jobHunter.connect();
	console.log(`✅ ${jobHunterConfig.name} (${jobHunterConfig.llmProvider}) initialized`);

	await cvCrafter.connect();
	console.log(`✅ ${cvCrafterConfig.name} (${cvCrafterConfig.llmProvider}) initialized`);

	await applicationTracker.connect();
	console.log(`✅ ${applicationTrackerConfig.name} (${applicationTrackerConfig.llmProvider}) initialized`);

	console.log(`\n✅ All 3 agents initialized!\n`);

	// Handle graceful shutdown
	process.on("SIGINT", async () => {
		console.log("\n\n🛑 Received SIGINT, shutting down gracefully...");
		await jobHunter.disconnect();
		await cvCrafter.disconnect();
		await applicationTracker.disconnect();
		process.exit(0);
	});

	process.on("SIGTERM", async () => {
		console.log("\n\n🛑 Received SIGTERM, shutting down gracefully...");
		await jobHunter.disconnect();
		await cvCrafter.disconnect();
		await applicationTracker.disconnect();
		process.exit(0);
	});

	// Run agents in a loop (every 5 minutes for Job Hunter, as needed for others)
	console.log("🚀 Starting agent iterations...\n");
	console.log("Press Ctrl+C to stop\n");

	let iteration = 0;
	const runIteration = async () => {
		iteration++;
		console.log(`\n${"=".repeat(60)}`);
		console.log(`ITERATION ${iteration}`);
		console.log(`${"=".repeat(60)}`);

		// Agent 1: Job Hunter (runs every iteration)
		await jobHunter.takeTurn(iteration);
		await new Promise((resolve) => setTimeout(resolve, 2000));

		// Agent 2: CV Crafter (runs every iteration)
		await cvCrafter.takeTurn(iteration);
		await new Promise((resolve) => setTimeout(resolve, 2000));

		// Agent 3: Application Tracker (runs every iteration)
		await applicationTracker.takeTurn(iteration);
		await new Promise((resolve) => setTimeout(resolve, 2000));

		console.log(`\n✅ Iteration ${iteration} complete!`);
		console.log(`⏰ Next iteration in 5 minutes...\n`);
	};

	// Run first iteration immediately
	await runIteration();

	// Then run every 5 minutes
	const interval = setInterval(async () => {
		await runIteration();
	}, 5 * 60 * 1000); // 5 minutes

	// Keep process alive
	process.on("SIGINT", () => {
		clearInterval(interval);
	});
}

main().catch((error) => {
	console.error("\n❌ Fatal error:", error);
	process.exit(1);
});

/**
 * Wait until the user has completed onboarding in the frontend.
 * We consider onboarding "ready" once the user profile exists with at least
 * one skill or experience entry so the agents have context to work with.
 */
async function waitForUserSetup(databaseUrl: string, userId: string) {
	const db = new JobDatabaseClient(databaseUrl);
	await db.connect();

	console.log("⏳ Waiting for user input from the frontend...");
	console.log("   Complete onboarding in the UI, then data will sync here.");

	while (true) {
		const profile = await db.getUserProfile(userId);
		const skills = profile?.skills || [];
		const experience = profile?.experience || [];

		if (profile && (skills.length > 0 || experience.length > 0)) {
			const applications = await db.getApplications(userId);
			const draftApplications = await db.getApplications(userId, "draft");
			const jobs = await db.getJobListings(5);

			await db.disconnect();

			return {
				profile,
				skillsCount: skills.length,
				experienceCount: experience.length,
				applications,
				draftApplications,
				jobListingsSample: jobs.slice(0, 3),
			};
		}

		console.log("   • Still waiting... no completed profile found. Retrying in 5s.");
		await new Promise((resolve) => setTimeout(resolve, 5000));
	}
}

/**
 * Print a friendly snapshot of the inputs before agents start iterating.
 */
function printUserSnapshot(snapshot: {
	profile: any;
	skillsCount: number;
	experienceCount: number;
	applications: any[];
	draftApplications: any[];
	jobListingsSample: any[];
}) {
	const { profile, skillsCount, experienceCount, applications, draftApplications, jobListingsSample } = snapshot;

	console.log("\n✅ User input detected — starting agents.");
	console.log("------------------------------------------------------------");
	console.log(`👤 Profile: ${profile.name || profile.user_id}`);
	console.log(`📍 Location: ${profile.location || "N/A"} | Target role: ${profile.target_role || "N/A"}`);
	console.log(`🧠 Skills: ${skillsCount} | Experience entries: ${experienceCount}`);
	console.log(`📄 Applications: ${applications.length} (draft: ${draftApplications.length})`);
	console.log("🗂️  Sample job listings the agents will see:");
	jobListingsSample.forEach((job) => {
		console.log(`   - ${job.title} at ${job.company} (${job.location || "Remote/Unknown"})`);
	});
	console.log("------------------------------------------------------------\n");
}
